Add jax.scipy.linalg.funm

ajcr opened the pull request:
This PR is to add `jax.scipy.linalg.funm`. It follows the same approach/algorithm as [`scipy.linalg.funm`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.funm.html) (including the same logic around dtype precision and utility methods).

ajcr committed:
Add jax.scipy.linalg.funm

ajcr commented:
Thanks for the review - I've put the utility methods that mirror the SciPy implementations in the third_party directory (I based this on the `signal_helper` module there).

The code that remains in `jax._src.scipy.linalg.py` still contains logic that is quite similar to that in SciPy (but expressed with JAX methods). For example, line 784 onwards where the correct `err` value is computed and printed. 

Do these parts of the implementation also need to go in the third party directory? Reading the license and code, I'm not quite sure where the line is.

jakevdp commented:
Any code derived from the scipy source should go in the third_party directory, even if it has been modified.

If you merely looked at the scipy function's API (i.e. which types of values it accepts and returns) and wrote your own implementation without referencing the scipy source code, then you can put it outside the third_party folder.

ajcr commented:
Makes sense - I've moved everything to the third_party directory.

On this topic, I should mention that I introduced `jax.scipy.linalg.rsf2csf` (which `funm` uses) in a [previous PR](https://github.com/google/jax/pull/10039) which was also based on the code in SciPy. I can move this to the third_party directory as well if that's appropriate.

jakevdp commented on jax/_src/third_party/scipy/linalg.py:
```diff
@@ -0,0 +1,127 @@
+  if jnp.isrealobj(A) and jnp.iscomplexobj(B):
+    if tol is None:
+      tol = {0: feps*1e3, 1: eps*1e6}[_array_precision[B.dtype.char]]
+    if jnp.allclose(B.imag, 0.0, atol=tol):
```
This will not be compatible with jax transforms, so we should avoid it. We need to statically determine the output type. This will mean breaking with the scipy API, but that's necessary to make the API JAX-compatible in this case.

There are other examples of this, but I can't recall where they are right now.

ajcr replied:
Thanks - I've removed the `_maybe_real` logic.

jakevdp commented on jax/_src/third_party/scipy/linalg.py:
```diff
@@ -0,0 +1,127 @@
+  if minden == 0.0:
+    minden = tol
+
+  if jnp.any(jnp.isinf(F)):
```
This is another statement that precludes transformation. Instead you can write it something like this:
```python
err = jnp.where(jnp.any(jnp.isinf(F)), jnp.inf, jnp.minimum(1, jnp.maximum(
            tol, (tol / minden) * norm(jnp.triu(T, 1), 1))))
```

jakevdp commented on jax/_src/third_party/scipy/linalg.py:
```diff
@@ -0,0 +1,127 @@
+            tol, (tol / minden) * norm(jnp.triu(T, 1), 1)))
+
+  if disp:
+    if err > 1000*tol:
```
Again, we cannot branch on non-static values, so we should remove this block.

jakevdp commented on tests/linalg_test.py:
```diff
@@ -1480,6 +1480,19 @@ def testRsf2csf(self, shape, dtype):
+  def testFunm(self, shape, dtype):
+      rng = jtu.rand_default(self.rng())
+      args_maker = lambda: [rng(shape, dtype), lambda x: x**-2.718]
+      self._CheckAgainstNumpy(osp.linalg.funm, jsp.linalg.funm, args_maker)
```
We should also call `_CompileAndCheck` once the above issues are addressed.

ajcr replied:
I think I've now added this check in correctly (I followed the tests for `jnp.apply_along_axis` for inspiration).

jakevdp reviewed (changes requested):
Looks good! There will be some work required here to make the implementation fully-compatible with JAX, but I think it's pretty close.

ajcr commented:
Thanks for the detailed feedback - I didn't fully appreciate at first that a JAX implementation of `funm` would be able to be compiled with its `func` argument (and that we'd be able to diverge from the SciPy API).

I've removed the `_maybe_real` call; as you say, the type of the output can't be determined through static analysis. I've added a note (`lax_description`) to the docstring to explain this and the fact that the error won't be printed when `disp=True`.

jakevdp commented on jax/_src/third_party/scipy/linalg.py:
```diff
@@ -0,0 +1,96 @@
+eps = jnp.finfo(float).eps
+feps = jnp.finfo(jnp.single).eps
+
+def _asarray_square(A):
```
Since this is used only once, and is fairly simple, maybe we should inline it?

jakevdp commented on jax/_src/third_party/scipy/linalg.py:
```diff
@@ -0,0 +1,96 @@
+return a complex-valued array.
+
+Additionally, unlike the SciPy implementation, when ``disp=True`` no warning
+will be printed if the error in the array output is estimated to be large.
```
We could possibly use the experimental [checkify](https://jax.readthedocs.io/en/latest/_autosummary/jax.experimental.checkify.check.html) work to implement something like this, but I'd probably leave that as a TODO.

jakevdp commented on jax/_src/third_party/scipy/linalg.py:
```diff
@@ -0,0 +1,96 @@
+  if disp:
+    return F
+
+  tol = {0: feps, 1: eps}[_array_precision[F.dtype.char]]
```
Won't this error for float16 and bfloat16?

Also, this seems needlessly abstracted. How about:
```
if F.dtype.char.lower() == 'f':
  tol = np.finfo(np.float32).eps
else:
  tol = np.finfo(np.float64).eps
```
with maybe another block for float16 types.

Having this logic defined in one place makes it much easier to understand the intent and effect of this line.

ajcr replied:
Agreed - I followed the SciPy source too closely here and the abstraction is less useful in this JAX code.

jakevdp commented on jax/scipy/linalg.py:
```diff
@@ -41,3 +41,5 @@
   polar_unitary as polar_unitary,
 )
+
+from jax._src.third_party.scipy.linalg import funm
```
We need to `import funm as funm` otherwise it will be treated as a private method by type checkers.

jakevdp commented on tests/linalg_test.py:
```diff
@@ -1480,6 +1480,32 @@ def testRsf2csf(self, shape, dtype):
+            "_shape={}_disp={}".format(jtu.format_shape_dtype_string(shape, dtype), disp),
+            "shape": shape, "dtype": dtype, "disp": disp
+        } for shape in [(1, 1), (5, 5), (20, 20), (50, 50)]
+            for dtype in float_types + complex_types
```
Keep all the `for` lines at the same indentation

jakevdp commented on tests/linalg_test.py:
```diff
@@ -1480,6 +1480,32 @@ def testRsf2csf(self, shape, dtype):
+        } for shape in [(1, 1), (5, 5), (20, 20), (50, 50)]
+            for dtype in float_types + complex_types
+              for disp in [True, False]))
+  @jtu.skip_on_devices("gpu", "tpu")
```
Why skip on gpu and tpu? Do we not expect this function to work correctly on those backends? If so we should mention this in the documentation. If not, we should test it.

ajcr replied:
`funm` uses `jax.scipy.linalg.schur` and my understanding is that Schur decomposition is currently implemented only for a CPU backend.

This also affects other methods such as `jax.scipy.linalg.sqrtm` which currently rely on Schur decomposition (and could affect yet-to-be-implemented methods such as `logm` if the same algorithm as SciPy is followed). 

I can add a note to the docstring explaining this (perhaps we'd want something generic that can be attached to other methods that may rely on Schur decomposition?).

jakevdp replied:
Makes sense. It would be worth adding a comment to that effect in the line where you skip the tests.

jakevdp commented on tests/linalg_test.py:
```diff
@@ -1480,6 +1480,32 @@ def testRsf2csf(self, shape, dtype):
+        scp_fun = lambda arr: osp.linalg.funm(arr, func, disp=disp)
+      else:
+        # if disp is False then funm returns a tuple (array, error) and
+        # the error is ignored here given it is likely to differ
```
We should not leave the error untested, particularly since it's not a straightforward computation. Why is it likely to differ? Can we test the value with an appropriate tolerance to make sure we're not returning entirely the wrong thing?

ajcr replied:
In my initial experimentatio, I thought I noticed a large discrepancy. I've checked again and the discrepancy is less than I feared - I'll make sure the test here checks the error as well as the array.

jakevdp commented on tests/linalg_test.py:
```diff
@@ -1480,6 +1480,32 @@ def testRsf2csf(self, shape, dtype):
+        # the error is ignored here given it is likely to differ
+        jnp_fun = lambda arr: jsp.linalg.funm(arr, func, disp=disp)[0]
+        scp_fun = lambda arr: osp.linalg.funm(arr, func, disp=disp)[0]
+      tol = 1e-06 if shape[0] < 50 else 1e-05
```
Setting tol this way overrides the default, which adjusts tol based on the dtype of the output. See other examples where `tol` is set to a dict mapping dtype to tolerance.

jakevdp commented on tests/linalg_test.py:
```diff
@@ -1480,6 +1480,32 @@ def testRsf2csf(self, shape, dtype):
+        scp_fun = lambda arr: osp.linalg.funm(arr, func, disp=disp)[0]
+      tol = 1e-06 if shape[0] < 50 else 1e-05
+      self._CheckAgainstNumpy(jnp_fun, scp_fun, args_maker, check_dtypes=False, tol=tol)
+      self._CompileAndCheck(jnp_fun, args_maker, check_cache_misses=False)
```
It would be best if we could pass this test with cache checks. To do this, the inner functions of the implementation would have to be defined at top level, rather than as closures.

jakevdp reviewed (changes requested):
Thanks for the updates - it's coming along! A few more comments below.

jakevdp commented:
The readthedocs failure is due to being based on an old `main` branch can you update `main` and rebase?

ajcr commented:
I think I've addressed the most recent set of comments and rebased on main.

However there's one issue I'm still looking into which is puzzling me. I'm checking `funm` on a `float16` dtype array and I consistently see errors like the following:
```python
import jax
import numpy as np

a = np.array([[3,2,1],[4,2,1],[4,4,0]], dtype=np.float16)

jax.scipy.linalg.funm(a, np.exp)
```
results in
```
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
malloc_consolidate(): invalid chunk size
Aborted (core dumped)
```
If I add a `np.float16` test case and run the unit tests, I see
```
[...]
[ RUN      ] ScipyLinalgTest.testFunm_shape=float16[1,1]_disp=False
[  FAILED  ] ScipyLinalgTest.testFunm_shape=float16[1,1]_disp=False
[ RUN      ] ScipyLinalgTest.testFunm_shape=float16[20,20]_disp=True
corrupted size vs. prev_size
Fatal Python error: Aborted

Current thread 0x00007f6a7b9dc740 (most recent call first):
[...call stack for thread...]
Aborted (core dumped)
```
Perhaps this is just my local machine/setup at fault, but I'm trying understand what's causing it.

jakevdp commented:
Can you tell if the float16 error is coming from the JAX function or from the scipy function?

ajcr commented:
I think it might `jax.scipy.linalg.schur` that's problematic on `np.float16` dtypes (at least on my laptop).
```
Python 3.10.4 (main, Mar 24 2022, 16:14:25) [GCC 7.5.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> import jax
>>> a = np.array([[3,2,1],[4,2,1],[4,4,0]], dtype=np.float16)
>>> jax.scipy.linalg.schur(a)
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
[DeviceArray([[-6.062e+00,  0.000e+00,  0.000e+00],
             [ 5.817e-05,  1.150e-05,  0.000e+00],
             [ 0.000e+00,  0.000e+00,  0.000e+00]], dtype=float16), DeviceArray([[0.   , 0.   , 0.   ],
             [0.   , 0.   , 1.984],
             [0.   , 0.   , 0.   ]], dtype=float16)]
>>> jax.scipy.linalg.schur(a)
free(): invalid next size (fast)
Aborted (core dumped)
```
(That Schur decomposition returned before the crash is very different to `scipy.linalg.schur(a)`.)

I've run the same code online in a colab notebook and it returned:
```
NotImplementedError: Unsupported dtype float16
```
(JAX version 0.3.4)

jakevdp commented:
Yeah, looks like a bug in the `schur` implementation for half-precision.

For now, I'd skip `float16` and `bfloat16` in the tests; I'll open a separate issue to report the schur error.

jakevdp commented:
The schur fix is in, though the fix is to raise a `NotImplementedError` rather than segfault. So for now you should limit the tests here to standard float & complex types.

jakevdp reviewed (approved):
Looks good!
